text/0000-float-semantics.md

```diff
@@ -0,0 +1,282 @@
+
+### Alternative: don't make any guarantee about the signaling/quiet bit ever
+
+This alternative would simplify the spec and make old MIPS hardware compliant.
+The signaling/quiet distinction also basically does not matter since floating-point exception flags are not exposed.
+However, there is one operation in C that can sometimes produce a non-NaN result on *quiet* NaN inputs specifically: [`pow`](https://en.cppreference.com/w/c/numeric/math/pow). For instance, "`pow(+1, exponent)` returns `1` for any `exponent`, even when `exponent` is `NaN`".
+C also says "This specification does not define the behavior of signaling NaNs", and in practice, `pow(1, sNaN)` returns a NaN.
+In other words, `pow` *does* [make a difference](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5a0855c2cb6630256e4407624650c4c9) between quiet and signaling NaN.
+So, if we say that any operation can arbitrairly produce signaling NaNs, then it becomes impossible to rely on `pow`'s property that "`pow(+1, exponent)` returns `1` for any `exponent`".
+Since the quiet/signaling distinction matters, we provide a guarantee which ensures that programs will almost always only ever deal with quiet NaNs (the only safe way to get a signaling NaN is to use `from_bits`).
+
+### Alternative: no strict semantic guarantees
+
+Providing strict IEEE 754-2008 guarantees precludes many transformations, such as turning `a*b + c` into FMA operations (since the result can change due to different rounding).
+We could provide weaker guarantees to allow such transformations.
+(That would also make using the higher-precision x87 instructions conforming with the spec.)
+However, there is no way to bound the effect of such transformations on program behavior: they lead to different outcome observable via `to_bits` and even via pure float operations, which can balloon into arbitrary changes in behavior.
+There is also no "monotonicity of precision": while e.g. an FMA instead of mul-then-add will lead to a higher-precision result for this particular operation, this can in principle lead to a lower-precision result in later computations.
+
+Given all these caveats, it seems preerrable to have explicit opt-in for such semantics-alterating transformations, e.g. via "fast" floating point operation intrinsics that don't provide strict IEEE semantics (and a type that uses those intrinsics for its operations).
+
+## Alternative options for `const` semantics
+
+### Alternative: `const` must be deterministic, and match runtime behavior
+
+This RFC proposes that we accept, for the first time, that a `const fn` can behave non-deterministically at run-time, and produce target/version/flag-specific results at compile-time. It can also produce different results when called at compile-time and at run-time.
+[Another proposed RFC](https://github.com/rust-lang/rfcs/pull/3352) gathers some general arguments for why we should allow such behavior in a `const fn`.
```

Comment:
Does this mean that if this RFC is merged, the other one vam be closed as this implies the other one? I think it makes more sense or accept the other one first instead of doing this semantic change here as a sidenote in this RFC.

Reply:
For the other RFC we were told we need a concrete instance  of some behavior we want to add to const-eval, we shouldn't just discuss and decide such things in the abstract. So I don't think we can land the other one first.

We could split this up, make this RFC only about runtime float semantics and then once it lands make a second RFC about compiletime float semantics. But I thought it makes more sense to discuss float semantics in general. I don't view this as just a sidenote in this RFC.